jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/Content.java

```diff
@@ -571,61 +574,67 @@ static Chunk next(Chunk chunk)
 
         /**
          * <p>Returns a new {@code Chunk} whose {@code ByteBuffer} is a slice of the
-         * {@code ByteBuffer} of the source {@code Chunk} unless the source
-         * {@link #hasRemaining() has no remaining byte} in which case:</p>
+         * whole {@code ByteBuffer} of the source {@code Chunk} unless the source
+         * {@link #hasRemaining() has no remaining bytes} in which case:</p>
          * <ul>
-         * <li>{@code this} is returned if it is an instance of {@link Error}</li>
-         * <li>{@link #EOF} is returned if {@link #isLast()} is {@code true}</li>
+         * <li>{@code this} is returned if {@link #isLast()} is {@code true}</li>
          * <li>{@link #EMPTY} is returned if {@link #isLast()} is {@code false}</li>
          * </ul>
-         * <p>If the source has remaining bytes, the returned {@code Chunk} retains
-         * the source {@code Chunk} and it is linked to it via
+         * <p>If the source {@code Chunk} has remaining bytes, the returned
+         * {@code Chunk} it is linked to the source {@code Chunk} via
          * {@link #from(ByteBuffer, boolean, Retainable)}.</p>
          *
-         * @return a new {@code Chunk} retained from the source {@code Chunk} with a slice
-         * of the source {@code Chunk}'s {@code ByteBuffer}
+         * @return a new {@code Chunk} with a slice of the source {@code Chunk}'s
+         * {@code ByteBuffer}
          */
         default Chunk slice()
         {
-            if (isTerminal())
-                return this;
-            if (!hasRemaining())
-                return EMPTY;
-            retain();
-            return from(getByteBuffer().slice(), isLast(), this);
+            if (hasRemaining())
+                return from(getByteBuffer().slice(), isLast(), this);
+            else
+                return isLast() ? this : EMPTY;
```

Comment:
The `from` method already has logic to check for remaining and do EOF/EMPTY return, so we should not duplicate here.   We also should retain here as a slice results in 2 chunks pointing to the same buffer.   The only place this method is currently called from is `ResponseNotifier#onChunk` and it does the retain anyway!

```suggestion
            retain();
            return from(getByteBuffer().slice(), isLast(), this);
```

Reply:
The point of not retaining in `slice()` is that it does not know how the returned slice is used.
Even if the (currently) only call stores the slice away for later use (and so has to retain), in future that may not be true, and we only change the code that actually uses the slice because that is what would have changed.